overlord/snapstate/handlers.go

```diff
@@ -1290,12 +1368,23 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	pb := NewTaskProgressAdapterLocked(t)
+
+	// if we have more revisions of this snap available, then restore the
+	// revision config and save the disabled services (which are restored when
+	// we go to link another snap)
 	if len(snapst.Sequence) > 0 {
+		// save the disabled services for later
+		err = m.saveSnapDisabledServices(snapst, pb)
```

Comment:
We're undoing, so we probably don't really care about what is disabled in the current revision of the snap and want to go back to the state of the old revision. 
Instead of querying systemd again, we could perhaps keep the old list of disabled services in the task and just restore that back into snapstate? I mean something like:

```
var oldDisabledServices []string
if err := t.Get("old-disabled-services", &oldDisabledServices); err != nil && err != state.ErrNoState {
    return err
}
snapst.LastActiveDisabledServices = oldDisabledServices
```

Reply:
well perhaps we could do it that way but it seems confusing and unnecessarily complicated that for undo specifically we hold onto those disabled services in the task, but then also hold onto those disabled services in the snap state. I think since we already need to do it in the state to handle cases like disable and later re-enable, we should only use that.

Furthermore, we do care what is disabled in the current revision since we are here, that means we already successfully completed a `link-snap` somewhere, and thus also called RestoreDisabledServices and deleted services which still exist in the new revision from the list in state, so if we don't save the disabled services here we would lose that state. Here's a concrete example of the situation where we need to save this for the link-snap:

1. snap needs a refresh
2. we figure out which services are disabled, save those in state
3. we unlink the current revision of the snap
4. we link the next revision of the snap, creating the systemd service units
5. we restore the disabled services, deleting the ones we disabled from the list saved in state
6. we then go to run the post-refresh hook, it dies and we need to undo the refresh
7. since linking the next revision in step 4 succeeded, we call the undo-link handler
8. now in the undo-link-snap handler, we already deleted from state the services we disabled, so the only way to know which services were disabled is by asking systemd.

Reply:
Now herein also lies _even more_ dragons :dragon: .... What if the post-refresh hook that failed above actually went and re-enabled the services that we disabled before it failed? Since the post-refresh hook doesn't queue service starts/stops/disables/enables, it could have deleted that state and so now when we roll back, we lost the state of those services that were disabled and so after a failed refresh, we now have additional services enabled which should have been left disabled. I think this case warrants also queuing service commands to be run after the hook successfully finishes, much like we do today for the configure hook.

Reply:
postponing further in hooks needs a discussion though, we cannot just do it in the name of bug fixing

Reply:
Also notice that having old-* values and use them for restoring is a pattern all over this code

Reply:
> postponing further in hooks needs a discussion though, we cannot just do it in the name of bug fixing

Sure we don't need to do it now, we can discuss more in depth before doing that, just something I noticed here